### PR TITLE
add tighter restrictions for the graduation year, fix #161 

### LINKED ISSUE #161: Users Inputting Wrong Graduation Dates
We should not allow graduation year of +5 current year (could be just client-side).

A real user:
image

Never trust our users to do the right things.

### PR DESCRIPTION
**Pull request by theTisch21**
This PR:
Edits the tests to be strict about which years are allowed for graduation
Edits the tests to dynamically update depending on the current year
Edits the validation to validate the graduation year against the new 5 year limit

### CODE REVIEW

**Comment by theTisch21** [comment]
This pr closes #161

**Comment by Zjjc123** [comment]
I haven't tested it but your formatting seems to be rejected by prettier checks I'm not sure why

**Comment by theTisch21** [comment]
Weird, I had set it to auto-format with prettier on save. I'll take a look at that tomorrow morning.

**Comment by theTisch21** [comment]
Huh, prettier didn't automatically run like I thought it did. In the future, I'll make sure to manually run prettier before committing. I've added a commit that fixes the formatting.

**Review by Zjjc123 — COMMENTED** [review]
Looks good the only thing is we can probably give user better feedback than just invalid hour. Like either tell them "Input numbers only" or like "Year can only be between [current year] and [current +5 years]

File: website/src/components/validation/validation.ts [hunk 1]
```diff
@@ -36,11 +36,23 @@ const VALIDATE_free_form = (text?: string): ValidationMessage => {
 };
 
 const VALIDATE_graduation = (text: string): ValidationMessage => {
-  const re = new RegExp("^(20)\\d\\d$");
-  if (!re.test(text)) {
-    return { validate: false, message: "Invalid Year" };
+  let thisyear = new Date().getFullYear();
```
Comment (Zjjc123) [inline website/src/components/validation/validation.ts:39]: also for these variables that you aren't changing just use const

File: website/src/components/validation/validation.ts [hunk 2]
```diff
@@ -36,11 +36,23 @@ const VALIDATE_free_form = (text?: string): ValidationMessage => {
...
+    //If this is true, the extracted number is a different length than the string, meaning there's at least one character not part of the number in the string
+    if (matches[0].length !== text.length) {
+      return { validate: false, message: "Invalid Year" };
+    }
+    //This grabs the actual number from the string
+    let num: number = parseInt(matches[0]);
+    //If the year submitted is either this year, or less than or equal to 5 years from now, it's valid
+    if (num >= thisyear && num <= thisyear + 5) {
+      return { validate: true, message: "" };
+    }
```
Comment (Zjjc123) [inline website/src/components/validation/validation.ts:52]:
```suggestion
    }
    else 
    {
       return { validate: false, message: "Graduation year should be between .... " };
    }
```

**Review by Zjjc123 — CHANGES_REQUESTED** [review]
But functionality-wise works great

**Review by Zjjc123 — DISMISSED** [review]

**Review by daniel-sudz — COMMENTED** [review]
Add floating points and  ```Nan``` to unit tests

**Review by daniel-sudz — CHANGES_REQUESTED** [review]
You can make your logic a whole of a lot easier to read without regex (merge it manually):

```js
const thisyear: number = new Date().getFullYear();
const gradYear: number = Number(text);
if (!Number.isInteger(gradYear)) {
  return { validate: false, message: "Invalid Year" };
} else if (gradYear >= thisyear && gradYear <= thisyear + 5) {
  return { validate: true, message: "" };
} else {
  return {
    validate: false,
    message: `Year should be between ${thisyear} and ${thisyear + 5}`,
  };
```

**Comment by theTisch21** [comment]
@daniel-sudz 
That is a lot simpler! Thanks!

File: website/src/components/validation/validation.test.ts [hunk 3]
```diff
@@ -37,18 +37,41 @@ test("Running Validation Tests for VALIDATE_free_form", () => {
...
+
+  //Tests graduation years slightly around the limit
+  expect(VALIDATE_graduation(thisYearPlus(-1)).validate).toBe(false);
+  expect(VALIDATE_graduation(thisYearPlus(6)).validate).toBe(false);
+  expect(VALIDATE_graduation(thisYearPlus(7)).validate).toBe(false);
+  expect(VALIDATE_graduation(thisYearPlus(10)).validate).toBe(false);
+  expect(VALIDATE_graduation(thisYearPlus(20)).validate).toBe(false);
 
+  //Tests bonkers years that should never be valid
   expect(VALIDATE_graduation("").validate).toBe(false);
```
Comment (daniel-sudz) [inline website/src/components/validation/validation.test.ts:64]:
```suggestion
  expect(VALIDATE_graduation("").validate).toBe(false);
  expect(VALIDATE_graduation(NaN).validate).toBe(false);
  expect(VALIDATE_graduation(2021.24).validate).toBe(false); //floating point
```

more possible edge cases

**Review by daniel-sudz — CHANGES_REQUESTED** [review]

**Comment by daniel-sudz** [comment]
@theTisch21 for the test cases, you need to fix the graduation year. Otherwise, the tests will start to fail a year from now lol. You can do this by adding an optional parameter ```currentYear``` to the validation and passing it in the test cases but not on the prod code.

File: website/src/components/validation/validation.test.ts [hunk 4]
```diff
@@ -37,18 +37,43 @@ test("Running Validation Tests for VALIDATE_free_form", () => {
   ).toBe(false);
 });
 
+//This function is here to avoid hard coded years in the tests, by returning the current year plus a specified number of years as a string
+function thisYearPlus(years: number): string {
+  let year = new Date().getFullYear();
+  year += years;
+  return year.toString();
+}
```
Comment (theTisch21) [inline website/src/components/validation/validation.test.ts:45]: > @theTisch21 for the test cases, you need to fix the graduation year. Otherwise, the tests will start to fail a year from now lol. You can do this by adding an optional parameter `currentYear` to the validation and passing it in the test cases but not on the prod code.
@daniel-sudz 
I'm not sure what needs fixing. This function (Lines 41-45 in the tests file) is designed to make the tests work years from now, and in the validation file, it takes the current year and compares the submitted year to that (Lines 38-51)

Comment (daniel-sudz) [reply website/src/components/validation/validation.test.ts:45]: Oh yeah nevermind I forgot that you switched it to be relative with ```thisYearPlus``` it should be good forever then pretty much

**Review by daniel-sudz — APPROVED** [review]
lgtm
